deep-learning/src/main/scala/com/microsoft/azure/synapse/ml/onnx/ONNXUtils.scala

```diff
@@ -0,0 +1,275 @@
+        OnnxTensor.createTensor(env, buffer, shape)
+      case OnnxJavaType.INT64 =>
+        val buffer = LongBuffer.allocate(size)
+        writeNestedSeqToBuffer[Long](batchedValues, buffer.put(_))
+        buffer.rewind()
+        OnnxTensor.createTensor(env, buffer, shape)
+      case OnnxJavaType.STRING =>
+        val flattened = writeNestedSeqToStringBuffer(batchedValues, size).toArray
+        OnnxTensor.createTensor(env, flattened, shape)
+      case other =>
+        throw new NotImplementedError(s"Tensor input type $other not supported. " +
+          s"Only FLOAT, DOUBLE, INT8, INT16, INT32, INT64, STRING types are supported.")
+    }
+  }
+
+  private def writeNestedSeqToBuffer[T: ClassTag](nestedSeq: Seq[_], bufferWrite: T => Unit): Unit = {
+    nestedSeq.foreach {
+      case x: T => bufferWrite(x)
+      case s: Seq[_] =>
+        writeNestedSeqToBuffer(s, bufferWrite)
+    }
+  }
+
+  private def writeNestedSeqToStringBuffer(nestedSeq: Seq[_], size: Int): ArrayBuffer[String] = {
+    var i = 0
+    val buffer = ArrayBuffer.fill[String](size)("")
+
+    def innerWrite(nestedSeq: Seq[_]): Unit = {
+      nestedSeq.foreach {
+        case x: String =>
+          buffer.update(i, x)
+          i = i + 1
+        case s: Seq[_] =>
+          innerWrite(s)
+      }
+    }
+
+    innerWrite(nestedSeq)
+    buffer
+  }
+
+  /**
+    * Returns true if the two data types are compatible. They are compatible if they share the same "shape", and
+    * 1. The element types from both sides are numeric types, or
+    * 2. The element types from both sides are the same.
+    */
+  @tailrec
+  private[onnx] def compatible(from: DataType, to: DataType): Boolean = {
+    (from, to) match {
+      case (VectorType, right: ArrayType) =>
+        compatible(DoubleType, right.elementType)
+      case (left: ArrayType, right: ArrayType) =>
+        compatible(left.elementType, right.elementType)
+      case (_: NumericType, _: NumericType) => true
+      case (fromDataType, toDataType) => fromDataType == toDataType
+    }
+  }
+
+  /*
+   * Create a new ONNXModel from an existing model, but shrink to only include given outputs
+   */
+  private[onnx] def sliceModelAtOutputs(fullModel: ONNXModel, outputs: Array[String]): ONNXModel = {
+    val fullProtobufModel = ModelProto.parseFrom(fullModel.getModelPayload)
+
+    val (newNodes, newOutputs) = findUsedNodesForOutputs(fullProtobufModel, outputs)
+
+    // Make a new model with the reduced set of nodes
+    val slicedGraph = makeGraph(newNodes, newOutputs, fullProtobufModel.getGraph)
+    val slicedProtobufModel = makeModel(slicedGraph, fullProtobufModel)
+
+    // Return a new model with all the same parameters, but with a new protobuf model
+    fullModel
+      .copy(ParamMap.empty)
+      .setModelPayload(slicedProtobufModel.toByteArray)
+  }
+
+  private def findUsedNodesForOutputs(model: ModelProto, newOutputNames: Array[String])
+      : (Array[NodeProto], Array[ValueInfoProto]) = {
```

Comment:
nit: odd spacing

Reply:
what is the spacing style suggested?  In this case the method def needs to be 2 lines due to line length.

Reply:
put start of args (before model:) on new line